File: 0402_remove_k_digits/remove_k_digits.py
def remove_k_digits(num: str, k: int) -> str:
    """
    Time: O(n)
    Space: O(n)
    """
    num_stack = []
    for n in num:
        while k and num_stack and num_stack[-1] > n:
            num_stack.pop()
            k -= 1
        num_stack.append(n)
    final = num_stack[:-k] if k else num_stack
    return "".join(final).lstrip("0") or "0"

File: 0402_remove_k_digits/remove_k_digits.go
package main

import "strconv"

/**
 * Time: O(n)
 * Space: O(n)
 */
func removeKdigits(num string, k int) string {
	numStack := make([]int, 1)
	for _, n := range num {
		for k != 0 && len(numStack) != 0 && numStack[len(numStack)-1] > int(n-'0') {
			k--
			numStack = numStack[:len(numStack)-1]
		}
		numStack = append(numStack, int(n-'0'))
	}
	for k != 0 {
		numStack = numStack[:len(numStack)-1]
		k--
	}
	res := ""
	flag := false
	for _, i := range numStack {
		if i > 0 {
			flag = true
		}
		if flag {
			res += strconv.Itoa(i)
		}
	}
	if res == "" {
		return "0"
	}
	return res
}

func main() {

}
